tensorflow-ops/src/TensorFlow/Gradient.hs

```diff
@@ -245,16 +253,17 @@ nonEmpty = anon mempty null
 -- | Calculate the gradients for every node in a graph.
 graphGrads :: forall a. GradientCompatible a
            => Graph
+           -> IntSet.IntSet
            -> Map FGL.Node (PendingGradients a)
            -- ^ Initial gradients (usually just 1 for the node of interest).
            -> Build (Map FGL.Node (Gradients a))
-graphGrads gr initPending = view gradientsResult <$> foldlM go initState nodeOrder
+graphGrads gr reachableSet initPending = view gradientsResult <$> foldlM go initState nodeOrder
   where
     initState = GradientsState initPending Map.empty
     -- Reverse topological sort.
-    -- TODO(fmayle): Filter out nodes that are not successors of any x in xs to
+    -- Filter out nodes that are not successors of any x in xs to
     -- avoid calculating gradients that won't be used.
-    nodeOrder = FGL.topsort $ FGL.grev gr
+    nodeOrder = filter (`IntSet.member` reachableSet) . FGL.topsort . FGL.grev $ gr
```

Comment:
I think there is a bit of an off by one error. If `xnodes` are in `reachableSet`, then we'll try to compute the gradients of the inputs to the nodes that output the `xs` tensors. That is probably why you couldn't delete the gradient functions for the rest of the undifferentiable ops.

One idea is to remove `xnodes` from `reachableSet` and then make the make the `reachableSet` membership check inside of `go`, after finalizing `outputGrads`, but before computing `inputGrads`.

```haskell
    nodeOrder = FGL.topsort $ FGL.grev gr
    go :: GradientsState a -> Int -> Build (GradientsState a)
    go state node = do
        -- Aggregate the accumulated gradients for this node.
        outputGrads <-
                sumPendingGradient (state ^. gradientsPending . at node . nonEmpty)
        if null outputGrads
           then pure state
           else do
              let nextState = state & gradientsResult %~ Map.insert node outputGrads
              if node `IntSet.member` reachableSet
                then do
                  let ctx = FGL.context gr node
                  inputGrads <- calculateInputGrads ctx outputGrads gr
                  -- Calculate the gradients for each of the node's inputs.
                  pure $ updatePendingGradients ctx inputGrads nextState
                else
                  pure nextState
```

(sorry the suggested code is ugly, my haskell skills are rusting)

Reply:
Thank you for the feedback and suggestions! This has been implemented and the `testEmbeddingLookupGradients` is fixed.